object_store/src/client/retry.rs

```diff
@@ -190,70 +198,108 @@ impl RetryExt for reqwest::RequestBuilder {
 
             loop {
                 let s = req.try_clone().expect("request body must be cloneable");
+
+                info!("The request being made: {:?}", s);
+
                 match client.execute(s).await {
-                    Ok(r) => match r.error_for_status_ref() {
-                        Ok(_) if r.status().is_success() => return Ok(r),
-                        Ok(r) if r.status() == StatusCode::NOT_MODIFIED => {
-                            return Err(Error::Client {
-                                body: None,
-                                status: StatusCode::NOT_MODIFIED,
-                            })
-                        }
-                        Ok(r) => {
-                            let is_bare_redirect = r.status().is_redirection() && !r.headers().contains_key(LOCATION);
-                            return match is_bare_redirect {
-                                true => Err(Error::BareRedirect),
-                                // Not actually sure if this is reachable, but here for completeness
-                                false => Err(Error::Client {
+                    Ok(r) => {
+                        info!("Response status: {}", r.status());
+                        info!("Response headers: {:#?}", r.headers());
+                        info!("Final URL of response: {}", r.url());
+    
+                        match r.error_for_status_ref() {
+                            // Response body might contain an Error despite the status saying 200.
+                            // More info here: https://repost.aws/knowledge-center/s3-resolve-200-internalerror
+                            Ok(r) if r.status().is_success() && r.text().await.contains("Error") => {
```

Comment:
better to only call `r.text().await` once. so you would have a block guarded by `if r.status().is_success()`, then within that block, call text to check for error